lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -699,25 +698,8 @@ sub _prepare_job_results ($self, $all_jobs, $limit) {
         running => 0,
         unknown => 0
     };
-    my $preferred_machines = _calculate_preferred_machines($all_jobs);
-
-    # read parameter for additional filtering
-    my $failed_modules = $self->param_hash('failed_modules');
-    my $states = $self->param_hash('state');
-    my $results = $self->param_hash('result');
-    my $archs = $self->param_hash('arch');
-    my $machines = $self->param_hash('machine');
-
-    my @jobs = grep {
-              (not $states or $states->{$_->state})
-          and (not $results or $results->{$_->result})
-          and (not $archs or $archs->{$_->ARCH})
-          and (not $machines or $machines->{$_->MACHINE})
-          and (not $failed_modules or $_->result eq OpenQA::Jobs::Constants::FAILED)
-    } @$all_jobs;
-    my $limit_exceeded = @jobs >= $limit;
-    @jobs = @jobs[0 .. ($limit - 1)] if $limit_exceeded;
-    my @jobids = map { $_->id } @jobs;
+    my @jobs = $jobs->all;
+    my $preferred_machines = _calculate_preferred_machines(\@jobs);
```

Comment:
just a note. The `$jobs` are used here only for calculate the `$preferred_machines` which is used later for this condition
``` if (   $job->MACHINE
            && $preferred_machines->{$job->ARCH}
            && $preferred_machines->{$job->ARCH} ne $job->MACHINE)```

So the `_prepare_job_results` needs this only and not the whole ResultSet. If you call `_calculate_preferred_machines(\@jobs);` before https://github.com/Martchus/openQA/blob/79f52e8c934da96eaa20e95fca56a8891dfc361a/lib/OpenQA/WebAPI/Controller/Test.pm#L848C5-L848C94 and just pass this. I think that way, it would simplify the dependencies between the invocations and the functions would be more clear. 

But lets not change anything now.

Reply:
> The $jobs are used here only for calculate …

The "only" in this sentence makes it wrong. Just search for `@jobs` in the subsequent code of this function; you'll find 4 occurrences.